File: src/main/java/com/example/probation/service/impl/CustomUserDetailsService.java
package com.example.probation.service.impl;

import com.example.probation.exception.ForbiddenException;
import com.example.probation.repository.UsersRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional
@RequiredArgsConstructor
public class CustomUserDetailsService implements UserDetailsService {
    private final UsersRepository usersRepository;

    @Override
    public UserDetails loadUserByUsername(final String username) {
        return usersRepository.findByUsername(username).orElseThrow(() ->
                new UsernameNotFoundException("User not found"));
    }

    public String getCurrentUsername() {
        final var auth = SecurityContextHolder.getContext().getAuthentication();
        final var anonymous = "anonymousUser";

        if (auth == null || anonymous.equals(auth.getName())) {
            throw new ForbiddenException();
        }

        return auth.getName();
    }
}

File: src/main/kotlin/com/example/probation/service/impl/CustomUserDetailsService.kt
package com.example.probation.service.impl

import com.example.probation.core.dto.UserDetailsDto
import com.example.probation.event.OnLoggingCompleteEvent
import com.example.probation.exception.ForbiddenException
import com.example.probation.exception.UserNotFoundException
import com.example.probation.repository.UsersRepository
import org.springframework.context.ApplicationEventPublisher
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.stereotype.Service
import javax.transaction.Transactional

@Service
@Transactional
class CustomUserDetailsService(
    private val usersRepository: UsersRepository,
    private val eventPublisher: ApplicationEventPublisher
) : UserDetailsService {
    companion object {
        const val ANONYMOUS = "anonymousUser"
    }

    override fun loadUserByUsername(username: String): UserDetails =
        usersRepository.findByUsername(username)?.let {
            UserDetailsDto(
                it.id,
                it.username,
                it.password,
                it.roles,
                it.enabled
            ).also { userDto ->
                eventPublisher.publishEvent(OnLoggingCompleteEvent(userDto))
            }
        } ?: throw UserNotFoundException("No such user, $username")

    fun getCurrentUsername(): String =
        SecurityContextHolder.getContext().authentication?.run {
            if (ANONYMOUS == name) {
                throw ForbiddenException("{forbidden}")
            }
            name
        } ?: throw ForbiddenException("{forbidden}")
}
